src/perf/best-practices.md

```diff
@@ -0,0 +1,411 @@
+since they allow Flutter to short-circuit most of the rebuild work.
+To create reusable pieces of UIs,
+prefer using a [`StatelessWidget`][] rather than a function.
+
+For more information, check out:
+
+* [Performance considerations][],
+  part of the [`StatefulWidget`][] API doc
+* [Widgets vs helper methods][],
+  a video from the official Flutter YouTube 
+  channel that explains why widgets
+  (especially widgets with `const` constructors)
+  are more performant than functions.
+
+[Performance considerations]: {{site.api}}/flutter/widgets/StatefulWidget-class.html#performance-considerations
+[source code for `SlideTransition`]: {{site.repo.flutter}}/blob/master/packages/flutter/lib/src/widgets/transitions.dart#L168
+[`StatefulWidget`]: {{site.api}}/flutter/widgets/StatefulWidget-class.html
+[`StatelessWidget`]: {{site.api}}/flutter/widgets/StatelessWidget-class.html
+[`TransitionBuilder`]: {{site.api}}/flutter/widgets/TransitionBuilder.html
+[Widgets vs helper methods]: {{site.youtube-site}}/watch?v=IOyq-eTRhvo
+
+---
+
+## Use saveLayer() thoughtfully
+
+Some Flutter code uses `saveLayer()`, an expensive operation,
+to implement various visual effects in the UI.
+Even if your code doesn't explicitly call `saveLayer()`,
+other widgets or packages that you use might call it behind the scenes.
+Perhaps your app is calling `saveLayer()` more than necessary.
+
+{{site.alert.secondary}}
+  <h4>Why is saveLayer expensive?</h4><a name="why-is-savelayer-expensive"></a>
+  Calling `saveLayer()` allocates an offscreen buffer.
+  Drawing content into the offscreen buffer might trigger
+  render target switches that are particularly slow in
+  older GPUs.
+{{site.alert.end}}
+
+How can you tell how much your app calls `saveLayer()`,
+either directly or indirectly?
+The `saveLayer()` method triggers
+an event on the [DevTools timeline][].
+If the calls to `saveLayer()` are excessive,
+this can cause jank.
+
+[DevTools timeline]: {{site.url}}/development/tools/devtools/performance#timeline-events-chart
+
+You have several options to handle this situation:
+
+* If the calls are coming from _your_ code, can you
+  reduce or eliminate them? For example, perhaps your
+  UI overlaps two shapes, each having non-zero transparency.
+  If they always overlap in the same amount, in the same
+  way, with the same transparency,
+  you can precalculate what this overlapped,
+  semi-transparent object looks like, cache it,
+  and use that instead of calling `saveLayer()`. 
+  This works with any static shape that you can precalculate.
+{% comment %}
+TBD: It would be nice if we could link to an example.
+  Kenzie suggested to John and Tao that we add an
+  example to perf_diagnosis_demo.)
+{% endcomment %}
+
+* If the calls are coming from a package that you don't own,
+  contact the package owner and ask why
+  these calls are necessary? Can they be reduced or
+  eliminated? If not, you might need to find another
+  package, or write your own.
```

Comment:
Applying some pressure to the owner might be good. It would be good to have guidance for owners here, like providing documentation of when a saveLayer might be necessary for their package and how to avoid it/when you cannot avoid it.

Reply:
ok, please double check my edits